internal/initializer/webhook_tls.go

```diff
@@ -0,0 +1,177 @@
+// updates the given secret.
+func (wt *WebhookCertificateGenerator) Run(ctx context.Context, kube client.Client) error {
+	s := &corev1.Secret{}
+	if err := kube.Get(ctx, wt.SecretRef, s); err != nil {
+		return errors.Wrap(err, errGetWebhookSecret)
+	}
+	// NOTE(muvaf): After 10 years, user will have to delete either of these
+	// keys and re-create the pod to have the initializer re-generate the
+	// certificate. No expiration check is put here.
+	if len(s.Data["tls.key"]) != 0 && len(s.Data["tls.crt"]) != 0 {
+		wt.log.Info("Given tls secret is already filled, skipping tls certificate generation")
+		return nil
+	}
+	wt.log.Info("Given tls secret is empty, generating a new tls certificate")
+	key, crt, err := wt.certificate.Generate(fmt.Sprintf("*.%s.svc", wt.ServiceNamespace))
+	if err != nil {
+		return errors.Wrap(err, errGenerateCertificate)
+	}
+	if s.Data == nil {
+		s.Data = make(map[string][]byte, 2)
+	}
+	s.Data["tls.key"] = key
+	s.Data["tls.crt"] = crt
+
+	return errors.Wrap(kube.Update(ctx, s), errUpdateWebhookSecret)
+}
+
+// NewRootCAGenerator returns a new RootCAGenerator.
+func NewRootCAGenerator() *RootCAGenerator {
+	return &RootCAGenerator{}
+}
+
+// RootCAGenerator generates a root CA and key that can be used by client and
+// servers.
+type RootCAGenerator struct{}
+
+// Generate creates TLS Secret with 10 years expiration date that is valid
+// for the given domains.
+func (*RootCAGenerator) Generate(domains ...string) (key []byte, crt []byte, err error) {
+	ca := &x509.Certificate{
+		SerialNumber: big.NewInt(2022),
+		Subject: pkix.Name{
+			CommonName:   "Crossplane",
+			Organization: []string{"Crossplane"},
+			Country:      []string{"Earth"},
+			Province:     []string{"Earth"},
+			Locality:     []string{"Earth"},
+		},
+		Issuer: pkix.Name{
+			CommonName:   "Crossplane",
+			Organization: []string{"Crossplane"},
+			Country:      []string{"Earth"},
+			Province:     []string{"Earth"},
+			Locality:     []string{"Earth"},
+		},
+		DNSNames:              domains,
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
+		IsCA:                  true,
+		KeyUsage:              x509.KeyUsageCRLSign | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+	// NOTE(muvaf): Why 2048 and not 4096? Mainly performance.
+	// See https://www.fastly.com/blog/key-size-for-tls
+	caPrivateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "cannot generate private key")
+	}
+	caBytes, err := x509.CreateCertificate(rand.Reader, ca, ca, &caPrivateKey.PublicKey, caPrivateKey)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "cannot create certificate with key")
+	}
+
+	caPEM := new(bytes.Buffer)
+	if err := pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caBytes,
+	}); err != nil {
+		return nil, nil, errors.Wrap(err, "cannot encode cert into PEM")
```

Comment:
It's not tested. It's quite hard to test this and even if you do, it takes a few seconds in each run so I tried to consolidate all RSA logic here and abstract it via an interface so that we can write tests for the tools that use this.